src/main/java/io/pinecone/configs/PineconeConnection.java

```diff
@@ -116,17 +135,18 @@ private static Metadata assembleMetadata(PineconeClientConfig clientConfig) {
         return metadata;
     }
 
-    static String getEndpoint(PineconeClientConfig clientConfig, PineconeConnectionConfig connectionConfig) {
-        String endpoint = (connectionConfig.getConnectionUrl() != null) ?
-                connectionConfig.getConnectionUrl().replaceFirst("https?://", "") :
-                String.format("%s-%s.svc.%s.pinecone.io",
-                        connectionConfig.getIndexName(),
-                        clientConfig.getProjectName(),
-                        clientConfig.getEnvironment());
-
-        logger.debug("Pinecone endpoint is: " + endpoint);
+    public static String getEndpoint(String host) {
+        if(host != null && !host.isEmpty()) {
+            return host.replaceFirst("https?://", "");
+        }
+        else {
+            throw new PineconeValidationException("Index host cannot be null or empty");
+        }
+    }
 
-        return endpoint;
+    private static String getHost(String apiKey, String indexName) {
+        PineconeControlPlaneClient controlPlaneClient = new PineconeControlPlaneClient(apiKey);
+        return controlPlaneClient.describeIndex(indexName).getHost();
```

Comment:
I feel like this may need to be a bit more robust in terms of possible failure of the `describeIndex` call. If that call throws, it may not be apparent to the user why it happened if we're intrinsically trying to resolve the host when they instantiated `PineconeConnection`.

Have you thought about having a cache that tracks index host URLs? We do this in the other clients with a singleton to prevent unnecessary calls to `describeIndex` when we've already resolved a specific host value. I think something like that backing this class may be useful as we could skip the calls to `describeIndex` if we've already retrieved it elsewhere.

Reply:
As we discussed it internally, I'll handle this as a part of a ticket where I'm trying to decouple ManagedChannels with indexes.